lib/src/generate/watch_impl.dart

```diff
@@ -126,40 +127,53 @@ class WatchImpl {
       }
       assert(_nextBuildScheduled == false);
 
-      /// Remove any updates that were generated outputs or otherwise not
-      /// interesting.
-      var updatesToRemove = updatedInputs.keys.where(_shouldSkipInput).toList();
-      updatesToRemove.forEach(updatedInputs.remove);
-      if (updatedInputs.isEmpty && !force) {
+      /// Copy [updatedInputs] so that it doesn't get modified after this point.
+      /// Any further updates will be scheduled for the next build.
+      ///
+      /// Only copy the "interesting" outputs.
+      var updatedInputsCopy = <AssetId, ChangeType>{};
+      updatedInputs.forEach((input, changeType) {
+        if (_shouldSkipInput(input)) return;
+        updatedInputsCopy[input] = changeType;
+      });
+      updatedInputs.clear();
+      if (updatedInputsCopy.isEmpty && !force) {
         return;
       }
 
       _logger.info('Preparing for next build');
-      _logger.info('Clearing cache for invalidated assets');
-      void clearNodeAndDeps(AssetId id, ChangeType rootChangeType) {
+      Future clearNodeAndDeps(
+          AssetId id, AssetId primaryInput, ChangeType rootChangeType) async {
         var node = _assetGraph.get(id);
         if (node == null) return;
-        if (node is GeneratedAssetNode) {
-          node.needsUpdate = true;
-        }
         _assetCache.remove(id);
         for (var output in node.outputs) {
-          clearNodeAndDeps(output, rootChangeType);
+          await clearNodeAndDeps(output, primaryInput, rootChangeType);
         }
 
         /// For deletes, prune the graph.
-        if (rootChangeType == ChangeType.REMOVE) {
+        if (id == primaryInput && rootChangeType == ChangeType.REMOVE) {
           _assetGraph.remove(id);
         }
+        if (node is GeneratedAssetNode) {
+          node.needsUpdate = true;
+          if (rootChangeType == ChangeType.REMOVE &&
+              node.primaryInput == primaryInput) {
+            _assetGraph.remove(id);
+            await _writer.delete(id);
+          }
+        }
       }
-      for (var input in updatedInputs.keys) {
-        clearNodeAndDeps(input, updatedInputs[input]);
-      }
-      updatedInputs.clear();
 
-      _logger.info('Starting build');
-      _currentBuild = _buildImpl.runBuild();
-      _currentBuild.then((result) {
+      () async {
```

Comment:
This still looks odd to me. If you prefer to leave the IIFE, please add a comment explaining why.